### ISSUE TITLE
need to explicitly catch RedisConnectionException to get the failover to work

### ISSUE THREAD

**Issue by josephnatale** [aterleto/jedis#7]
## Setup

* 2 clusters running locally
  * cluster 1  - locahost:6379
  * cluster 2 - localhost:6380
* the test application is just a unit test looping doing 1 write / second 
* to create the failure scenario I'm simply shutting down the primary cluster mid test

## Results

When I don't explicitly catch the RedisConnectionException the library does not fail over, my unit test fails
```
writing value 5
2023-05-25 19:22:42.400  WARN 66839 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T19:22:42.400598-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379', waiting PT0.5S until attempt '1'. Last attempt failed with exception 'redis.clients.jedis.exceptions.JedisConnectionException: Unexpected end of stream.'.
2023-05-25 19:22:42.909  WARN 66839 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T19:22:42.908953-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379', waiting PT1S until attempt '2'. Last attempt failed with exception 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'.
2023-05-25 19:22:43.916 ERROR 66839 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T19:22:43.916362-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379' recorded a failed retry attempt. Number of retry attempts: '3'. Giving up. Last exception was: 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'.
2023-05-25 19:22:43.917 ERROR 66839 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T19:22:43.917370-05:00[America/Chicago]: CircuitBreaker 'cluster:1:127.0.0.1:6379' recorded an error: 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'. Elapsed time: 1533 ms

Failed to connect to any host resolved for DNS name.
redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.
	at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccessfulHost(DefaultJedisSocketFactory.java:63)
	at redis.clients.jedis.DefaultJedisSocketFactory.createSocket(DefaultJedisSocketFactory.java:87)
	at redis.clients.jedis.Connection.connect(Connection.java:188)
	at redis.clients.jedis.Connection.initializeFromClientConfig(Connection.java:349)
	at redis.clients.jedis.Connection.<init>(Connection.java:61)
	at redis.clients.jedis.ConnectionFactory.makeObject(ConnectionFactory.java:68)
	at org.apache.commons.pool2.impl.GenericObjectPool.create(GenericObjectPool.java:571)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:298)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:223)
	at redis.clients.jedis.util.Pool.getResource(Pool.java:38)
	at redis.clients.jedis.ConnectionPool.getResource(ConnectionPool.java:29)
	at redis.clients.jedis.providers.MultiClusterPooledConnectionProvider$Cluster.getConnection(MultiClusterPooledConnectionProvider.java:305)
	at redis.clients.jedis.executors.CircuitBreakerCommandExecutor.handleExecuteCommand(CircuitBreakerCommandExecutor.java:59)
	at redis.clients.jedis.executors.CircuitBreakerCommandExecutor.lambda$executeCommand$0(CircuitBreakerCommandExecutor.java:45)
	at io.github.resilience4j.retry.Retry.lambda$decorateSupplier$2(Retry.java:213)
	at io.github.resilience4j.circuitbreaker.CircuitBreaker.lambda$decorateSupplier$4(CircuitBreaker.java:197)
	at io.github.resilience4j.core.SupplierUtils.lambda$recover$5(SupplierUtils.java:119)
	at redis.clients.jedis.executors.CircuitBreakerCommandExecutor.executeCommand(CircuitBreakerCommandExecutor.java:52)
	at redis.clients.jedis.UnifiedJedis.executeCommand(UnifiedJedis.java:183)
	at redis.clients.jedis.UnifiedJedis.set(UnifiedJedis.java:669)
	at com.discover.paymentservices.h2.redis.config.JedisFunctionalTests.testCircuitBreak(JedisFunctionalTests.java:32)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:135)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:829)
	Suppressed: java.net.ConnectException: Connection refused (Connection refused)
		at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
		at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:412)
		at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:255)
		at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:237)
		at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
		at java.base/java.net.Socket.connect(Socket.java:609)
		at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccessfulHost(DefaultJedisSocketFactory.java:73)
		... 110 more


JedisFunctionalTests > testCircuitBreak() FAILED
    redis.clients.jedis.exceptions.JedisConnectionException at JedisFunctionalTests.java:32
1 test completed, 1 failed
> Task :test FAILED
```

Based off the logs what I've gathered is that for each unifiedJedis.set(...) invocation the library will a retry the configured amount of times.   If based off the configuration the library determines in needs to fail over whatever exception that caused the failover (in this case RedisConnectionException) is thrown as a Runtime exception.  

## initial thoughts

* Appreciate this unit test isn't really a great represent of a production workloads but if I'm understanding the functionality right a single invocation of unifiedJedis.set(...) will result in some abandonded reads and writes.  As with issue #6 we are still in far better shape than before this capability existed
* Lightbulb just went of that issue #6 and this issue may be related....

**Comment by aterleto** [comment]
Hi Joe, everything is configurable i.e. #retries, which exceptions are registered, what type of sliding window to use (time or amount), how many failures would trigger a failover within the sliding window, etc. It can even track slow commands as your team initially requested, although this is hard to test because its Redis.

About losing data - let's look at an extreme case where we want to failover **after** the first connection issue. An example of this is the following unit test - https://github.com/aterleto/jedis/blob/cross-cluster-failover/src/test/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProviderTest.java#L84 - **testRunClusterFailoverPostProcessor**

In that test, I configure the sliding-window and min-calls within it to 1. This means that if 1 out of 1 commands were to generate a connection exception then **on the next subsequent command** it will close the circuit breaker and failover with success.
```
MultiClusterClientConfig.Builder builder = new MultiClusterClientConfig.Builder(clusterConfigs);

// Configures a single failed command to trigger an open circuit on the next subsequent failure
builder.circuitBreakerSlidingWindowSize(1);
builder.circuitBreakerSlidingWindowMinCalls(1);
```

So, as you correctly stated, if that initial failed command was not retried by the application (not to be confused with the internal retries) then it would be lost. However if the application was to retry then it would work on the subsequent operation.

**Comment by josephnatale** [comment]
Ok, cool to just to confirm the behavior...any failures that occur before the circuit opens need to be retried by the calling function, once the circuit opens failover will occur and subsequent requests will go down the next connection in the list. 

I went in with the (perhaps naive) assumption that if we got say a "RedisConnectionException" for a particular 'SET' or 'GET' that would be retried after the circuit opens against the the secondary connection.  **This is a clear case of a user not understanding how the product works** 😄 

I'm going to close this and #6 as its expected behavior.
